src/ncp/spinel_decoder.hpp

```diff
@@ -0,0 +1,523 @@
+ *    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ *    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ *    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file contains the definitions of a spinel decoder.
+ */
+
+#ifndef SPINEL_DECODER_HPP_
+#define SPINEL_DECODER_HPP_
+
+#include <openthread/config.h>
+
+#include <openthread/ncp.h>
+#include <openthread/types.h>
+
+#include "openthread-core-config.h"
+#include "ncp/spinel.h"
+
+namespace ot {
+namespace Ncp {
+
+/**
+ * This class defines a spinel decoder.
+ *
+ */
+class SpinelDecoder
+{
+public:
+
+    /**
+     * This constructor initializes a `SpinelDecoder` object
+     *
+     */
+    SpinelDecoder(void);
+
+    /**
+     * This method initializes the decoder to start decoding a new given spinel encoded frame.
+     *
+     * It sets the read position to beginning of the frame and also erases/voids any saved positions (see
+     * `SavePosition()` and `ResetToSaved()` methods).
+     *
+     * @param[in] aFrame                Pointer to the buffer containing the frame to be decoded.
+     * @param[in] aLength               Length (number of bytes) of the frame.
+     *
+     */
+    void Init(const uint8_t *aFrame, uint16_t aLength);
+
+    /**
+     * This method returns a pointer to the frame.
+     *
+     * @returns A pointer to buffer containing current frame being decoded.
+     *
+     */
+    const uint8_t *GetFrame(void) const { return mFrame; }
+
+    /**
+     * This method returns the total length of current frame being decoded.
+     *
+     * @returns The length of current frame being decoded.
+     *
+     */
+    uint16_t GetLength(void) const { return mLength; }
+
+    /**
+     * This method returns the number of bytes that are already read/decoded from the frame.
+     *
+     * @returns The number of bytes already read from frame.
+     *
+     */
+    uint16_t GetReadLength(void) const { return mIndex; }
+
+    /**
+     * This method returns the number of remaining (not yet read/decoded) bytes in the frame.
+     *
+     * @returns The number of remaining unread bytes in the frame.
+     *
+     */
+    uint16_t GetRemainingLength(void) const { return mLength - mIndex; }
+
+    /**
+     * This method indicates whether or not all the bytes in the frame are read.
+     *
+     * @returns TRUE if all the bytes in the buffer are read, FALSE otherwise.
+     *
+     */
+    bool IsAllRead(void) const { return (mIndex == mLength); }
+
+    /**
+     * This method resets the read position to beginning of frame. It will also void/erase any previously saved
+     * position using `SavePosition()` method.
+     *
+     */
+    void Reset(void);
+
+    /**
+     * This method decodes and reads a boolean value form the frame. On success, the read position is updated.
```

Comment:
nit: first line should be 1-line summary.  Empty line below 1-line summary.  Then additional detail as necessary follows (applies to doxygen below as well).

Reply:
Thanks. Fixed them in the file...